A36-xds-for-servers.md

```diff
@@ -54,57 +54,71 @@ also allow it access to the server to change behavior (e.g., via an
 interceptor). Users will use the XdsServer API to construct the server instead
 of the existing API.
 
-Credential configuration will be managed separately by providing a
-XdsServerCredential, similar to client-side. The user must pass the
-XdsServerCredential to the XdsServer API to enable control-plane managed TLS
-certificates. The user is free to pass other ServerCredential types to XdsServer
-and they will be honored. Note that RBAC is authz, not authn, so it would be
-enabled by using the XdsServer, even without XdsServerCredential.
+Credential configuration will be managed separately by providing
+XdsServerCredentials, similar to client-side. The user must pass the
+XdsServerCredentials to the XdsServer API to enable control-plane managed TLS
+certificates. While the user-facing API is discussed here, the precise
+implementation details are covered in gRFC A29. The user is free to pass other
+ServerCredentials types to XdsServer and they will be honored. Note that RBAC is
+authz, not authn, so it would be enabled by using the XdsServer, even without
+XdsServerCredentials.
 
 When starting the XdsServer, it will start communication with xDS to receive
-initial configuration. `bind()` and `listen()` can happen immediately, but
-`accept()`ing on the port will be delayed until the initial configuration is
+initial configuration. `bind()` and `listen()` may happen immediately, but
```

Comment:
I think we should actually write down all of the considerations we've discussed about this, so that future readers understand the implications.  Here's some possible wording; feel free to change as needed:

"""
The server should ideally avoid calling `listen()` on the port until after it has obtained the initial configuration, so that clients attempting to connect to the server will quickly fail, thus preserving reasonable pick_first behavior.  If for some reason that is not feasible, a reasonable alternative would be to start listening but immediately `accept()` and then `close()` each incoming connection attempt, since this will result in essentially the same behavior on the client.  A less desirable alternative would be to call `listen()` and then simply not call `accept()` on incoming connection attempts, since that may cause the connection attempts to timeout instead of immediately failing, depending on the size of the accept queue.

The server must call `bind()` before calling `listen()`, but some implementations cannot make those two calls at different times.  However, if the implementation desires to support ephemeral ports, it must call `bind()` to determine the actual port being used before it can request the xDS resource.  Therefore, server implementations that cannot call `bind()` and `listen()` separately must choose between (a) not supporting ephemeral ports and calling `bind()` and `listen()` after receiving the xDS resource and (b) supporting ephemeral ports by calling `bind()` and `listen()` before receiving the xDS resource, using the `accept()` and `close()` approach.  (Note that we do not expect that ephemeral ports are likely to be a common case with xDS in the first place.)

Note that the above considerations about when to call `bind()` and `listen()` arise not only when the server first starts up but also when the server receives an xDS update indicating that the required resource has been removed.  Although the same considerations apply to both cases, implementations may choose different trade-offs in each case.  For example, an implementation may choose at startup to defer calling `listen()` until it receives the xDS resource, but when it receives an update telling it that the xDS resource has been removed, it may be easier to `accept()` and `close()` incoming connections rather than to stop listening.
"""

I realize that some of this is currently covered in the "Rationale" section, but I think it's a little easier to understand if all of the considerations are described in one place.  I think the "Rationale" section can basically just point out that an alternative way of ameliorating the client pick_first behavior would be to implement Happy Eyeballs on the client side; it won't need to cover anything else, because the rest will be covered here.

Reply:
As per discussion in https://github.com/grpc/grpc/pull/24956#discussion_r541417034, we should also document the fact that when the Listener resource is removed by the xDS server, we will also close any pre-existing connections.

CC @yashykt